File: WinRTCompWRLItf/Citizen.cpp
// Citizen.cpp
#include "pch.h"
#include "Citizen.h"
#include "Address.h"

using namespace WinRTCompWRLItf;
using namespace Platform;

Citizen::Citizen()
{
	m_address = ref new WinRTCompWRLItf::Address(); 
}

String^ Citizen::Name::get()
{
	return ref new String(m_name.c_str());
}

void Citizen::Name::set(String^ value)
{
	m_name = value->Data();
}

String^ Citizen::Surname::get()
{
	return ref new String(m_surname.c_str());
}

void Citizen::Surname::set(String^ value)
{
	m_surname = value->Data();
}

IAddress^ Citizen::Address::get()
{
	return m_address;
}

void Citizen::Address::set(IAddress^ value)
{
	m_address = value;
}

bool Citizen::CanSave()
{
	return 
		m_name.length() > 0 && 
		m_surname.length() > 0 && 
		((ISaveable^) m_address)->CanSave();
}

File: WinRTCompCsharpWRLItf/Citizen.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

using WRLCompV1;

namespace WinRTCompCsharpWRLItf
{
    /// <summary>
    /// The class Citizen implements in C# the following interfaces of WRLCompV1 namespace:
    ///     - ICitizen
    ///     - IPerson
    ///     - ISaveable
    /// </summary>
    public sealed class Citizen : ICitizen, IPerson, ISaveable
    {
        private IPerson person;
        private IAddress address;

        public Citizen()
        {
            person = new Person();
            address = new Address();
        }

        #region Explicit implementation of IPerson

        string IPerson.Name
        {
            get
            {
                return person.Name;
            }
            set
            {
                person.Name = value;
            }
        }

        string IPerson.Surname
        {
            get
            {
                return person.Surname;
            }
            set
            {
                person.Surname = value;
            }
        }

        #endregion

        #region Explicit implementation of ICitizen

        IAddress ICitizen.Address
        {
            get
            {
                return address;
            }
            set
            {
                address = value;
            }
        }

        string ICitizen.Name
        {
            get
            {
                return ((IPerson) this).Name;
            }
            set
            {
                ((IPerson)this).Name = value;
            }
        }

        string ICitizen.Surname
        {
            get
            {
                return ((IPerson)this).Surname;
            }
            set
            {
                ((IPerson)this).Surname = value; ;
            }
        }

        #endregion

        #region Explicit implementation of ISaveable

        public bool CanSave()
        {
            return
                ((ISaveable)person).CanSave() &&
                ((ISaveable)address).CanSave();
        }

        #endregion
    }
}
